### ISSUE TITLE
Bench rigor: force device sync inside timed region of nature_dqn CNN GPU benches

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/thrust#335]
## Context

Discovered during Judge review of PR #334 (issue #328, Epic #306 Phase 3 large-net crossover benchmarks).

The four new `nature_dqn_*` criterion groups in `benches/trainer_throughput.rs` time closures that return **device-resident tensors / models** without ever forcing a host read-back or an explicit backend sync:

- `bench_nature_dqn_policy_forward` / `bench_nature_dqn_qnet_forward` return raw `Tensor<B, _>` from `black_box(...)` — no `.into_data()` / sync.
- `bench_nature_dqn_policy_train_step` / `bench_nature_dqn_qnet_train_step` return the updated model — no host read.

This contrasts with the existing MLP `*_train_step` groups, whose timed closures return a host-side `stats` struct (which materializes scalars back to host, incidentally forcing a sync).

## Why it is not a correctness blocker for #334

On async GPU backends (wgpu/Metal, cuda) this could in principle under-report GPU time. However, the PR's measured numbers show strong evidence that real compute is being timed, not just kernel enqueue:

- The numbers scale with batch size (`policy_forward` 1.88 ms @ b32 → 13.1 ms @ b256; `qnet_forward` 5.73 → 19.5 ms). Pure enqueue-only timing would be ~batch-independent.
- Criterion runs hundreds of iterations per timed sample, so GPU command-queue backpressure bounds any under-report to ~queue-depth iterations (negligible).
- The numbers are physically plausible (~1 TFLOP/s effective on M3 Ultra; ~11 GFLOP/s single-threaded ndarray).

So the crossover verdict (GPU wins the CNN workload, decisively) is robust.

## What to improve

For maximum rigor and apples-to-apples parity with the real inference path (which always reads logits/actions back to host):

1. Add an explicit backend sync or a `.into_data()` read-back of a single element inside the timed closure of the `*_forward` groups.
2. Consider the same for the `*_train_step` groups if the optimizer step does not already force materialization.
3. Re-measure and, if the absolute wgpu/Metal numbers shift materially, refresh the `docs/BURN_BACKENDS.md` large-net table (the operator cuda run is already pending on alc-2, so this can ride along).

Low priority — the current numbers are directionally sound and honestly caveated.

**Comment by rjwalters** [comment]
Curating this issue per user request (filed by judge during PR #334 review, no labels yet).

## Curator Enhancement

### Problem Statement

The four `nature_dqn_*` criterion groups added in PR #334 (`bench_nature_dqn_policy_forward`, `bench_nature_dqn_qnet_forward`, `bench_nature_dqn_policy_train_step`, `bench_nature_dqn_qnet_train_step` in `benches/trainer_throughput.rs`) have timed closures that return device-resident values without triggering a host read-back or an explicit backend sync:

- `bench_nature_dqn_policy_forward`: `black_box((logits, values))` — two device tensors
- `bench_nature_dqn_qnet_forward`: `black_box(q_values)` — device tensor
- `bench_nature_dqn_policy_train_step`: `black_box(policy)` — updated model (all parameters on device)
- `bench_nature_dqn_qnet_train_step`: `black_box(qnet)` — updated model (all parameters on device)

On async GPU backends (wgpu/Metal, CUDA) this can allow GPU work to be deferred past the point where criterion stops the clock — potentially under-reporting the true compute time.

**Contrast with existing MLP benchmarks.** The eight original groups (`a2c_train_step`, `ppo_train_step`, `dqn_train_step`, `sac_train_step`, and their full-loop siblings) return host-scalar `f64` stats structs (`A2cStats`, `DQNStepStatsBurn`, `SacStepStats`) computed inside the trainer via `.into_scalar()` / `.into_data()` calls, which force a GPU-to-host transfer and thereby act as an implicit sync. The CNN groups have no equivalent materialization.

**Why it is not a correctness blocker (judge's evidence from PR #334 review).** The measured wgpu/Metal numbers scale with batch size — `nature_dqn_policy_forward` goes from 1.88 ms at B=32 to 13.1 ms at B=256 (~7× for 8× batch), and `nature_dqn_qnet_forward` from 5.73 ms to 19.5 ms (~3.4×). Pure kernel-enqueue-only timing would be near-batch-independent. Additionally, criterion runs hundreds of iterations per timed sample, so GPU command-queue backpressure bounds any under-report to approximately one queue-depth of iterations. The crossover verdict (GPU wins the CNN workload decisively) is robust in direction.

**Why the fix still matters.** For maximum rigor and apples-to-apples parity with the real inference/training path (which always reads logits/actions/loss scalars back to host), the timed closures should force a sync. This also ensures the numbers will remain correct if Burn's internal command queue depth changes in a future version or if a new async backend is added.

### Burn 0.21 Sync API

Burn 0.21's `Backend` trait (`burn::tensor::backend::Backend`) exposes:

```rust
fn sync(_device: &Self::Device) -> Result<(), ExecutionError> {
    Ok(())  // default: no-op (NdArray/CPU)
}
```

The `Autodiff<B>` wrapper (i.e., the `AutodiffBackend` type) delegates directly to `B::sync(device)`. The `burn-cubecl` implementation (which backs both `wgpu` and `cuda`) calls `futures_lite::future::block_on(client.sync())`, draining the command queue. The NdArray backend's default no-op means CPU numbers are semantically unaffected.

`B::sync(device)` is accessible in the bench file with the existing `AutodiffBackend` bound, provided `Backend` is in scope:

```rust
use burn::tensor::backend::Backend; // add to existing import block
```

The bench file already has `use burn::tensor::{..., backend::AutodiffBackend}` — adding `Backend` to that path is the only import change needed.

### Implementation Guidance

#### Sync mechanism: prefer `B::sync(device)` over `.into_scalar()`

Two mechanisms are available:

**Option A — `B::sync(device)` (recommended)**

```rust
// At the end of each timed closure, after black_box:
let _ = B::sync(device);
```

Pros: explicit intent, zero tensor allocation, no data transferred, identical cost on NdArray (no-op). Cons: needs `device` captured in the closure (already available via `iter_batched` setup pattern). Requires adding `use burn::tensor::backend::Backend` import.

**Option B — `.into_scalar()` on one output element**

```rust
// In policy_forward, after the forward call:
let (logits, values) = policy.forward(obs);
let _ = logits.flatten::<1>(0, 1).slice([0..1]).into_scalar(); // forces sync
black_box((logits, values))  // logits already consumed — would need clone
```

Pros: no additional import. Cons: requires cloning the tensor before the forced read (allocates), or restructuring the return to be after the sync — awkward. Also transfers data (one f32) unnecessarily.

**Option A is the right choice for all four groups.** It adds negligible overhead (nanoseconds on CPU, a fence drain on GPU — which is what we want to measure), requires one import change, and cleanly expresses the intent.

#### Per-closure prescription

**`bench_nature_dqn_policy_forward`** — add `B::sync(device)` after `black_box((logits, values))`. The `device` reference is available in the outer bench function and must be moved into the timed closure. Since `iter_batched` closures are `move` by default when capturing references, `device` needs to be captured as `&B::Device`:

```rust
|(policy, obs)| {
    let (logits, values) = policy.forward(obs);
    black_box((logits, values));
    let _ = B::sync(device);
},
```

**`bench_nature_dqn_qnet_forward`** — same pattern:

```rust
|(qnet, obs)| {
    let q_values = qnet.forward(obs);
    black_box(q_values);
    let _ = B::sync(device);
},
```

**`bench_nature_dqn_policy_train_step`** — the timed closure already has `opt` and `policy` in scope; `device` must be captured:

```rust
|(policy, mut opt, obs, actions)| {
    let (log_probs, entropy, values) = policy.evaluate_actions(obs, actions);
    let loss = log_probs.mean().neg() + values.powf_scalar(2.0).mean()
        - entropy.mean().mul_scalar(CNN_ENTROPY_COEF);
    let grads = GradientsParams::from_grads(loss.backward(), &policy);
    let policy = opt.inner_mut().step(CNN_LR, policy, grads);
    black_box(policy);
    let _ = B::sync(device);
},
```

**`bench_nature_dqn_qnet_train_step`** — same:

```rust
|(qnet, mut opt, obs)| {
    let q_values = qnet.forward(obs);
    let loss = q_values.powf_scalar(2.0).mean();
    let grads = GradientsParams::from_grads(loss.backward(), &qnet);
    let qnet = opt.inner_mut().step(CNN_LR, qnet, grads);
    black_box(qnet);
    let _ = B::sync(device);
},
```

#### Cost of the added sync

- **CPU / NdArray**: zero cost — `B::sync` is a no-op. Numbers are unaffected semantically.
- **wgpu / Metal / CUDA**: drains the command queue — a fence wait that in steady state adds one round-trip of host-GPU latency. On a loaded GPU this is typically 10–100 µs; for the CNN groups (which measure 2–500 ms of GPU work) this is well below 1% overhead. It does NOT change what is being measured — it ensures the measurement includes all GPU work, rather than risking early cutoff.

### Re-measurement Protocol

After the sync is in place, re-run CPU and wgpu locally with a shortened criterion window to check whether absolute numbers shift materially:

```bash
# CPU baseline
cargo bench --features training --bench trainer_throughput \
    -- --sample-size 10 --warm-up-time 2 --measurement-time 10 nature_dqn

# wgpu/Metal
cargo bench --features "training,wgpu" --bench trainer_throughput \
    -- --sample-size 10 --warm-up-time 2 --measurement-time 10 'nature_dqn.*wgpu'
```

**Definition of "materially":** any group whose median shifts by more than 10% relative to the values in the current `docs/BURN_BACKENDS.md` large-net table. If any group shifts materially, update that table and note the methodology change in the "Host setup / Command" block (one line: "Timed closures now include an explicit `B::sync(device)` call to drain the GPU command queue before criterion stops the clock").

**The cuda column does NOT need a re-run here.** That column is still a placeholder (pending the alc-2 operator run). The operator run after this fix lands will naturally use the corrected bench code. The doc's cuda command block should note that the in-region sync is present.

**Expected outcome:** numbers are unlikely to shift materially (the backpressure argument holds), but if they do, the refreshed values are more trustworthy. If they do not shift, a one-sentence methodology note in the docs is sufficient — no table update needed.

### Scope Guard

- Changes are limited to `benches/trainer_throughput.rs` (four timed closures + one `use` import) and possibly `docs/BURN_BACKENDS.md` (methodology note and/or number refresh if shift > 10%).
- No changes to `src/`, `Cargo.toml`, or any other file.
- The fix is backwards-compatible: the NdArray path is semantically unchanged (sync is a no-op).

## Affected Files

- `benches/trainer_throughput.rs` — add `use burn::tensor::backend::Backend;` import; add `let _ = B::sync(device);` at the end of the timed closure in each of the four `bench_nature_dqn_*` functions
- `docs/BURN_BACKENDS.md` — add a methodology note to the large-net "Host setup / Command" block stating that timed closures include an in-region `B::sync(device)` call; update the large-net table numbers if any group shifts >10% after re-measurement

## Test Plan

- [ ] Compile check (no sync): `cargo bench --features training --bench trainer_throughput --no-run` compiles cleanly
- [ ] Compile check (wgpu): `cargo bench --features "training,wgpu" --bench trainer_throughput --no-run` compiles cleanly
- [ ] Lint: `cargo clippy --all-targets --features training -- -D warnings` and `cargo clippy --all-targets --features "training,wgpu" -- -D warnings` both clean
- [ ] Format: `cargo fmt --check` clean
- [ ] CPU smoke run: `cargo bench --features training --bench trainer_throughput -- --sample-size 3 --warm-up-time 1 --measurement-time 3 nature_dqn` completes without panic; numbers are in the same order of magnitude as the current BURN_BACKENDS.md table
- [ ] wgpu smoke run: `cargo bench --features "training,wgpu" --bench trainer_throughput -- --sample-size 3 --warm-up-time 1 --measurement-time 3 'nature_dqn.*wgpu'` completes without panic on a GPU host
- [ ] Materiality check: compare CPU and wgpu medians against the current BURN_BACKENDS.md large-net table; if any group shifts >10%, update the table and add the methodology note; if no group shifts >10%, add the methodology note only (no table update needed)
- [ ] Existing groups unaffected: all eight original groups (`a2c_train_step`, etc.) still present and unmodified in `register_all`
